pkg/cloudmap/cache.go

```diff
@@ -0,0 +1,148 @@
+package cloudmap
+
+import (
+	"fmt"
+	"github.com/aws/aws-cloud-map-mcs-controller-for-k8s/pkg/model"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/util/cache"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"time"
+)
+
+const (
+	nsKeyPrefix    = "ns"
+	svcKeyPrefix   = "svc"
+	endptKeyPrefix = "endpt"
+
+	defaultCacheSize = 1024
+	defaultNsTTL     = 2 * time.Minute
+	defaultSvcTTL    = 2 * time.Minute
+	defaultEndptTTL  = 5 * time.Second
+)
+
+type ServiceDiscoveryClientCache interface {
+	GetNamespace(namespaceName string) (namespace *model.Namespace, found bool)
+	CacheNamespace(namespace *model.Namespace)
+	CacheNilNamespace(namespaceName string)
+	GetServiceId(namespaceName string, serviceName string) (serviceId string, found bool)
+	CacheServiceId(namespaceName string, serviceName string, serviceId string)
+	GetEndpoints(serviceId string) (endpoints []*model.Endpoint, found bool)
+	CacheEndpoints(serviceId string, endpoints []*model.Endpoint)
+	EvictEndpoints(serviceId string)
```

Comment:
I am just thinking out loud here. Why do we support `negative cache` for namespaces and not for other resource types? Does it make sense to treat services differently from namespaces?